### PR TITLE
Consolidate periodic loops into one function updating Den and updating autostop.

### PR DESCRIPTION
**Pull request by rohinb2**
autostop.

### CODE REVIEW

**Comment by rohinb2** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/run-house/runhouse/873?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#873** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/873?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#871** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/871?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rohinb2 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: runhouse/servers/cluster_servlet.py [hunk 1]
```diff
@@ -260,63 +246,45 @@ async def aclear_all_references_to_env_servlet_name(self, env_servlet_name: str)
     # Cluster status functions
     ##############################################
 
-    async def asend_status_info_to_den(self):
+    async def aperiodic_status_check(self):
         # Delay the start of post_status_thread, so we'll finish the cluster startup properly
+        print("hello")
         await asyncio.sleep(STATUS_CHECK_DELAY)
+        print("delayed")
```
Comment (dongreenberg) [inline runhouse/servers/cluster_servlet.py:253]: Ding

File: runhouse/rns/rns_client.py [hunk 2]
```diff
@@ -659,3 +671,25 @@ def contents(self, name_or_path, full_paths):
         return folder(name=name_or_path, path=folder_url).resources(
             full_paths=full_paths
         )
+
+    def send_status(self, status: ResourceStatusData, cluster_rns_address: str):
```
Comment (dongreenberg) [inline runhouse/rns/rns_client.py:675]: We need to catch all other errors so any failures here don't nuke the loop or cluster servlet.

Comment (dongreenberg) [reply runhouse/rns/rns_client.py:675]: Should this be async?

File: runhouse/servers/cluster_servlet.py [hunk 3]
```diff
@@ -260,63 +246,45 @@ async def aclear_all_references_to_env_servlet_name(self, env_servlet_name: str)
     # Cluster status functions
     ##############################################
 
-    async def asend_status_info_to_den(self):
+    async def aperiodic_status_check(self):
```
Comment (dongreenberg) [inline runhouse/servers/cluster_servlet.py:249]: Funny name

File: runhouse/servers/autostop_servlet.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import time
+
+
+class AutostopServlet:
+    """A helper class strictly to run SkyPilot methods on OnDemandClusters inside SkyPilot's conda env."""
+
+    def __init__(self):
+        self._last_activity = time.time()
+        self._last_register = None
```
Comment (dongreenberg) [inline runhouse/servers/autostop_servlet.py:9]: I don't think we need the last_registered last_activity stuff anymore. We can just set the activity directly in SkyPilot, there's no reason to set it lazily like we did before.

File: runhouse/servers/cluster_servlet.py [hunk 5]
```diff
@@ -85,10 +71,10 @@ async def __init__(
 
         # Only send for clusters that have den_auth enabled and if we are logged in with a user's token
         # to authenticate the request
-        if self.cluster_config.get("den_auth", False) and configs.token:
-            logger.debug("Creating send_status_info_to_den thread.")
+        if self.cluster_config.get("den_auth", False):
+            logger.info("Creating aperiodic_status_check thread.")
```
Comment (dongreenberg) [inline runhouse/servers/cluster_servlet.py:75]: Ok this has to be on purpose

**Review by dongreenberg — COMMENTED** [review]
Looks solid, could be simplified a little and some extra fault tolerance

File: runhouse/rns/rns_client.py (hunk 2 above)
Comment (rohinb2) [reply runhouse/rns/rns_client.py:675]: Guess async doesn't rly matter because its in its own thread, but can for consistency

Comment (rohinb2) [reply runhouse/rns/rns_client.py:675]: > We need to catch all other errors so any failures here don't nuke the loop or cluster servlet.

The cluster servlet has everything in a try: except: so it'll catch it

File: runhouse/servers/autostop_servlet.py (hunk 4 above)
Comment (rohinb2) [reply runhouse/servers/autostop_servlet.py:9]: Nah currently we also do `set_last_active_time_to_now` in `get_env_servlet_name_for_key`. This happens quite often, we don't need to re-set it everytime, and the loop is always running, so we can just do it there.

File: runhouse/servers/cluster_servlet.py (hunk 5 above)
Comment (rohinb2) [reply runhouse/servers/cluster_servlet.py:75]: haha

**Comment by rohinb2** [comment]
### Merge activity

* **Jun 7, 8:13 PM EDT**: @rohinb2 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/run-house/runhouse/873).

**Pull request merged**